src/org/etools/j1939_84/controllers/part03/Part03Step03Controller.java

```diff
@@ -53,13 +70,73 @@
-        // 6.3.3.3.a. Warn if (if supported) any ECU additional DTCs are provided than the DTC observed in step 6.3 in a positive DM27 response.
+        // 6.3.3.1.a Global DM27 (send Request (PGN 59904) for PGN 64898 (SPNs 1213-1215, 3038, 1706)).
+        List<DM27AllPendingDTCsPacket> globalPackets = getDiagnosticMessageModule().requestDM27(getListener())
+                .getPackets();
+        List<Integer> obdModuleAddresses = getDataRepository().getObdModuleAddresses();
+        obdModuleAddresses.forEach(this::clearLastDm27);
+
+        globalPackets.stream()
+                .peek(this::verifyDtcsAreSame)
+                .peek(this::checkForAdditionalDtc)
+                .forEach(this::updateObdLastDM27);
+
+        // 6.3.3.4.a DS DM27 to each OBD ECU.
+        List<RequestResult<DM27AllPendingDTCsPacket>> dsResults = obdModuleAddresses.stream()
+                .map(address -> getDiagnosticMessageModule()
+                        .requestDM27(getListener(), address)
+                        .requestResult())
+                .collect(Collectors.toList());
+        List<DM27AllPendingDTCsPacket> dsPackets = filterRequestResultPackets(dsResults);
 
-        // 6.3.3.4.a. DS DM27 to each OBD ECU.
         // 6.3.3.5.a Fail if (if supported) any difference compared to data received with global request.
-        // 6.3.3.5.a Fail if NACK not received from OBD ECUs that did not respond to global query.
+        compareRequestPackets(globalPackets, dsPackets, "6.3.3.5.a");
+
+        // 6.3.3.5.b Fail if NACK not received from OBD ECUs that did not respond to global query.
+        List<AcknowledgmentPacket> dsAcks = filterRequestResultAcks(dsResults);
+        checkForNACKs(globalPackets, dsAcks, obdModuleAddresses, "6.3.3.5.b");
+    }
+
+    private void clearLastDm27(int obdAddress){
+        OBDModuleInformation obdModule = getDataRepository().getObdModule(obdAddress);
+        obdModule.setLastDM27(null);
+        getDataRepository().putObdModule(obdModule);
     }
 
+    private void updateObdLastDM27(DM27AllPendingDTCsPacket packet){
+        OBDModuleInformation obdModule = getDataRepository().getObdModule(packet.getSourceAddress());
+        obdModule.setLastDM27(packet);
+        getDataRepository().putObdModule(obdModule);
+    }
+
+    private void verifyDtcsAreSame(DM27AllPendingDTCsPacket packet) {
+        List<DiagnosticTroubleCode> packetDtcs = packet.getDtcs();
+        OBDModuleInformation obdModule = getDataRepository().getObdModule(packet.getSourceAddress());
+        List<DiagnosticTroubleCode> obdDtcs = obdModule.getEmissionDTCs();
+        // 6.3.3.2.a Fail if (if supported) no ECU reports the same DTC observed in step 6.3.2.1 in a positive DM27 response.
+        if (packetDtcs.size() != obdDtcs.size() || !packetDtcs.equals(obdDtcs)) {
+            addFailure(
+                    "6.3.3.2.a - OBD module " + getAddressName(obdModule.getSourceAddress()) +
+                    " saved dtcs are:" + NL +
```

Comment:
Let's not going into the details of the errors.  It makes the report longer and makes our coding more complicated.  The user can compare the two values size by size to determine what the issue is.